pkg/image/registry/imagestreammapping/rest.go

```diff
@@ -90,18 +102,52 @@ func (s *REST) Create(ctx kapi.Context, obj runtime.Object) (runtime.Object, err
 		Image:                image.Name,
 	}
 
-	if !api.AddTagEventToImageStream(stream, tag, next) {
-		// nothing actually changed
-		return &unversioned.Status{Status: unversioned.StatusSuccess}, nil
-	}
+	var result *unversioned.Status
+	var updateConflictErr error
+	retryErr := client.RetryOnConflict(wait.Backoff{Steps: maxRetriesOnConflict}, func() error {
+		lastEvent := api.LatestTaggedImage(stream, tag)
+		if !api.AddTagEventToImageStream(stream, tag, next) {
+			// nothing actually changed
+			result = &unversioned.Status{Status: unversioned.StatusSuccess}
+			return nil
+		}
+		api.UpdateTrackingTags(stream, tag, next)
 
-	api.UpdateTrackingTags(stream, tag, next)
+		_, err := s.imageStreamRegistry.UpdateImageStreamStatus(ctx, stream)
+		if err != nil {
+			if !errors.IsConflict(err) {
+				return err
+			}
+			// If the update conflicts, get the latest stream and check for tag
```

Comment:
All of this code should be at the beginning of the loop - guarded by "don't do this on the first loop".  Before the call to RetryOnConflict, set the last stream tag.  At the beginning of the loop, check whether stream == nil.  Then check whether old tag == new tag.  If they don't, then exit.  Do the mutation.  If you get *any* error, clear stream (so the next iteration of the loop reads it again).